### PR TITLE
Add tracer of torchscript

### PR DESCRIPTION
**Pull request by junjihashimoto**
- [x] Trace simple adder
- [x] Trace parameterized model
- [x] Build on macos(Probably ghc treats input-file as c language)
- [x] Insert parameters into traced model

mlp's torchscript is below.

```python
op_version_set = 1
class M(Module):
  __parameters__ = []
  def forward(self: __torch__.M,
    argument_1: Tensor,
    argument_2: Tensor,
    argument_3: Tensor,
    argument_4: Tensor,
    argument_5: Tensor,
    argument_6: Tensor,
    argument_7: Tensor) -> Tensor:
    _0 = torch.addmm(argument_3, argument_1, torch.t(argument_2), beta=1, alpha=1)
    _1 = torch.addmm(argument_5, torch.relu(_0), torch.t(argument_4), beta=1, alpha=1)
    _2 = torch.addmm(argument_7, torch.relu(_1), torch.t(argument_6), beta=1, alpha=1)
    return torch.log_softmax(_2, 1, 6)
```

### CODE REVIEW

**Comment by junjihashimoto** [comment]
The code to decide the language is here.
https://github.com/ghc/ghc/blob/ghc-8.6.5-release/compiler/main/DriverPipeline.hs#L1222-L1233

**Comment by junjihashimoto** [comment]
https://github.com/fpco/inline-c/blob/master/inline-c-cpp/src/Language/C/Inline/Cpp.hs#L27
https://github.com/fpco/inline-c/blob/92c3ca8ecedc12471441336c5f6cc4cdd3f604f4/inline-c/src/Language/C/Inline/Internal.hs#L158

I do not know why there is not c++-file but c-file in /tmp directory.

**Comment by junjihashimoto** [comment]
@apaszke 
trace function of pr uses callback below.
TensorList of "A" should be the same as TensorList of "B", 
though I think neither callback-function and argument need to be forcing.

```haskell
trace :: (ForeignPtr TensorList{-A-} -> IO (ForeignPtr TensorList)) 
         -> ForeignPtr TensorList{-B-}
         -> IO (ForeignPtr Module)
trace traced_function arguments_of_traced_function = ..
```

**Comment by apaszke** [comment]
I don't mean that we have to force the callback nor the arguments. I've been thinking about the callback's outputs! It's possible that the casting routines already do that, but it would be good to make check that it really is the case.

**Comment by SurajK7** [comment]
Sorry if this is a bit off-topic. Can we convert a list of torch.nn.parameter.Parameter from PyTorch to a list of tensors in hasktorch? @junjihashimoto 
Sorry if the question is too naive I am still trying to get my head around hasktorch.

**Comment by junjihashimoto** [comment]
@SurajK7 
If torch.nn.parameter.Parameter you think is the same as libtorch's parameters, we can do it.
I tried following code. The code uses libtorch's Module and paramters.
vector<at:Tensor> is the same as hasktorch's [Tensor].
Then the code extracts vector<at:Tensor>  from Module' parameters.
https://gist.github.com/junjihashimoto/018ba1a7a31e833622423150fac2d7eb

File: hasktorch/test/ScriptSpec.hs [hunk 1]
```diff
@@ -1,32 +1,109 @@
...
     b `shouldBe` 2
-    (asValue c :: Float) `shouldBe` 4.0
+    (asValue c :: Float) `shouldBe` 8.0
+    -- save sm "self.pt"
+    -- sm <- load "self.pt"
+    -- let IVTuple [IVInt a,IVInt b,IVTensor c] = runMethod1 sm "forward" (IVTensor (ones' []))
+    -- let [g] =  grad c (flattenParameters sm)
+    -- (asValue g :: Float) `shouldBe` 1.0
+    -- -- This throws 'CppStdException "Exception: Differentiated tensor not require grad; type: std::runtime_error"'.
+    -- -- See libtorch-ffi/src/Torch/Internal/Unmanaged/Autograd.hs
```
Comment (junjihashimoto) [inline hasktorch/test/ScriptSpec.hs]: @apaszke 
This code throws a exception for '!roots[0].function == true' of Autograd.hs.
Do you think it is related to callback's outputs not forced?

File: libtorch-ffi/src/Torch/Internal/Unmanaged/Type/Module.hs [hunk 2]
```diff
@@ -85,27 +87,86 @@ train obj on = [C.throwBlock| void {
...
+    for(auto p : parameters) {
+      vec_parameters->push_back(p);
+    }
+    return vec_parameters;
+  }|]
+
+
+setParameters :: Ptr Module -> Ptr TensorList -> IO ()
+setParameters obj params = [C.throwBlock| void {
+    auto parameters = $(torch::jit::script::Module* obj)->parameters();
```
Comment (junjihashimoto) [inline libtorch-ffi/src/Torch/Internal/Unmanaged/Type/Module.hs]: This paremeters  are just copy.
https://github.com/pytorch/pytorch/blob/v1.4.1/torch/csrc/jit/script/module.h#L295

File: hasktorch/test/ScriptSpec.hs [hunk 3]
```diff
@@ -1,32 +1,131 @@
...
     a `shouldBe` 1
     b `shouldBe` 2
-    (asValue c :: Float) `shouldBe` 4.0
+    (asValue c :: Float) `shouldBe` 8.0
+    sm2@(UnsafeScriptModule sm2') <- load "self.pt"
+    p0 <- makeIndependent (asTensor (3::Float))
+    setParameters (UnsafeRawModule sm2') [(toDependent p0)]
+    let IVTuple [IVInt a,IVInt b,IVTensor c] = runMethod1 sm2 "forward" (IVTensor (ones' []))
+    let [g] =  grad c (flattenParameters sm2)
+    (asValue g :: Float) `shouldBe` 1.0
```
Comment (junjihashimoto) [inline hasktorch/test/ScriptSpec.hs]: We can get gradient from torchscript.
I am refactoring these codes.

File: hasktorch/hasktorch.cabal [hunk 4]
```diff
@@ -76,6 +76,8 @@ library
                     , safe-exceptions
                     , zlib >= 0.6
                     , inline-c
+ default-extensions:  Strict
+                    , StrictData
```
Comment (junjihashimoto) [inline hasktorch/hasktorch.cabal:80]: @apaszke 
I add Strict pragma to not use laziness.
https://gitlab.haskell.org/ghc/ghc/-/wikis/strict-pragma

**Comment by junjihashimoto** [comment]
All issues have been completed. (Supporting onnx is out of scope.)
